public/lib/validate.js

```diff
@@ -0,0 +1,142 @@
+   * @returns {number} - Current node index. Returns `-1` if node name not found.
+   */
+  const _getNodeId = (nodes, name) => nodes.findIndex(node => node.name === name);
+
+  /**
+   * Add node to node-list.
+   *
+   * @param {array<NODE_OBJ>} nodes - List of node objects.
+   * @param {string} name - The node name.
+   * @returns {array<NODE_OBJ>} - New list of node objects.
+   */
+  const _addNode = (nodes, name) => {
+    let id = _getNodeId(nodes, name);
+    if (id < 0) {
+      id = nodes.push({ name });
+    }
+    return nodes;
+  }
+
+  /**
+   * Add link to link-list.
+   *
+   * @param {array<NODE_OBJ>} nodes - List of node objects.
+   * @param {array<LINK_OBJ>} links - List of link objects.
+   * @param {LINK_OBJ} link - The link object. 
+   * @returns {array<LINK_OBJ>} - New list of link objects. 
+   */
+  const _addLink = (nodes, links, link) => {
+    links.push({
+      source: _getNodeId(nodes, link.source),
+      target: _getNodeId(nodes, link.target),
+      value: link.value
+    });
+    return links;
+  }
+
+  /**
+   * Cast hashmap to object with two attributes,
+   * list of nodes and list of links.
+   *
+   * @param {HASHMAP} data - The hashmap.
+   * @returns {object} - With all nodes and links. 
+   */
+  const _castHashMap2Object = (data) => {
+    let nodes = [];
+    let links = [];
+    data.forEach((link) => {
+      nodes = _addNode(nodes, link.source);
+      nodes = _addNode(nodes, link.target);
+      links = _addLink(nodes, links, link);
+    });
+    return {nodes, links};
+  }
+
+  /**
+   * Filters invalid values from hashmap.
+   * An invalid value could be a `0` value.
+   *
+   * @param {HASHMAP} data - The hashmap.
+   * @returns {HASHMAP} - Without invalid values.
+   */
+  const _filterHashMap = (data) => {
+    data.forEach((link, key) => {
+      if (!link.value) {
+        data.delete(key);
+      }
+    });
+    return data;
+  }
+
+  /**
+   * Generate a hashmap.
+   * Generates a key for each link.
+   *
+   * @param {array<NODE_OBJ>} nodes - List of node objects.
+   * @param {array<LINK_OBJ>} links - List of link objects.
+   * @returns {HASHMAP} - Hashmap of links.
+   */
+  const _generateHashMap = (nodes, links) => {
+    const data = new Map();
+    links.map((link, index) => {
+      data.set(index, {
+        source: _getNodeName(nodes, link.source),
+        target: _getNodeName(nodes, link.target),
+        value: link.value
+      });
+    });
+    return data;
+  }
+
+  /**
+   * Remove unvalid nodes or links.
+   * 
+   * @param {array<NODE_OBJ>} nodes - List of node objects.
+   * @param {array<LINK_OBJ>} links - List of link objects.
+   * @returns {object} - Valid nodes and links.
+   */
+  const validate = (nodes, links) => {
+    // generate hashmap
+    const dataMap = _generateHashMap(nodes, links);
+    // delete unvalid links and nodes
+    const filteredDataMap = _filterHashMap(dataMap);
+    // split hashmap into two arrays
```

Comment:
I would rephrase this comment to something like
convert linkMap to link and node array